File: python/iceberg/api/transforms/identity.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from .transform import Transform
from .transform_util import TransformUtil
from ..expressions import Expressions
from ..types import TypeID


class Identity(Transform):

    @staticmethod
    def get(type_var):
        return Identity(type_var)

    def __init__(self, type_var):
        self.type_var = type_var

    def apply(self, value):
        return value

    def can_transform(self, type_var):
        return type_var.is_primitive_type()

    def get_result_type(self, source_type):
        return source_type

    def project(self, name, predicate):
        return self.project_strict(name, predicate)

    def project_strict(self, name, predicate):
        if predicate.lit is not None:
            return Expressions.predicate(predicate.op, name, predicate.lit.value)
        else:
            return Expressions.predicate(predicate.op, name)

    def to_human_string(self, value):
        if value is None:
            return "null"

        if self.type_var.type_id == TypeID.DATE:
            return TransformUtil.human_day(value)
        elif self.type_var.type_id == TypeID.TIME:
            return TransformUtil.human_time(value)
        elif self.type_var.type_id == TypeID.TIMESTAMP:
            if self.type_var.adjust_to_utc:
                return TransformUtil.human_timestamp_with_timezone(value)
            else:
                return TransformUtil.human_timestamp_without_timezone(value)
        elif self.type_var.type_id in (TypeID.BINARY, TypeID.FIXED):
            raise NotImplementedError()
            # if isinstance(value, bytearray):
            #     return base64.b64encode(value)
            # elif isinstance(value, bytes):
            #     return base64.b64encode(bytes(value))
            # else:
            #     raise RuntimeError("Unsupported binary type: %s" % value.__class__.__name__)
        else:
            return str(value)

    def __str__(self):
        return "identity"

    def __eq__(self, other):
        if id(self) == id(other):
            return True

        if other is None or not isinstance(other, Identity):
            return False

        return self.type_var == other.type_var

    def __hash__(self):
        return hash(self.__key())

    def __key(self):
        return Identity.__class__, self.type_var

File: api/src/main/java/org/apache/iceberg/transforms/Identity.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.iceberg.transforms;

import java.io.ObjectStreamException;
import org.apache.iceberg.expressions.BoundPredicate;
import org.apache.iceberg.expressions.Expressions;
import org.apache.iceberg.expressions.UnboundPredicate;
import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
import org.apache.iceberg.types.Type;
import org.apache.iceberg.util.SerializableFunction;

class Identity<T> implements Transform<T, T> {
  private static final Identity<?> INSTANCE = new Identity<>();

  private final Type type;

  /**
   * Instantiates a new Identity Transform
   *
   * @deprecated use {@link #get()} instead; will be removed in 2.0.0
   */
  @Deprecated
  public static <I> Identity<I> get(Type type) {
    return new Identity<>(type);
  }

  @SuppressWarnings("unchecked")
  public static <I> Identity<I> get() {
    return (Identity<I>) INSTANCE;
  }

  private static class Apply<T> implements SerializableFunction<T, T> {
    private static final Apply<?> APPLY_INSTANCE = new Apply<>();

    @SuppressWarnings("unchecked")
    private static <T> Apply<T> get() {
      return (Apply<T>) APPLY_INSTANCE;
    }

    @Override
    public T apply(T t) {
      return t;
    }
  }

  private Identity() {
    this(null);
  }

  private Identity(Type type) {
    this.type = type;
  }

  @Override
  public T apply(T value) {
    return value;
  }

  @Override
  @SuppressWarnings("checkstyle:HiddenField")
  public SerializableFunction<T, T> bind(Type type) {
    Preconditions.checkArgument(canTransform(type), "Cannot bind to unsupported type: %s", type);
    return Apply.get();
  }

  @Override
  public boolean canTransform(Type maybePrimitive) {
    return maybePrimitive.isPrimitiveType();
  }

  /**
   * Returns a human-readable String representation of a transformed value.
   *
   * <p>null values will return "null"
   *
   * @param value a transformed value
   * @return a human-readable String representation of the value
   * @deprecated use {@link #toHumanString(Type, Object)} instead; will be removed in 2.0.0
   */
  @Deprecated
  @Override
  public String toHumanString(T value) {
    if (this.type != null) {
      return toHumanString(this.type, value);
    }
    return Transform.super.toHumanString(value);
  }

  @Override
  public Type getResultType(Type sourceType) {
    return sourceType;
  }

  @Override
  public boolean preservesOrder() {
    return true;
  }

  @Override
  public boolean satisfiesOrderOf(Transform<?, ?> other) {
    // ordering by value is the same as long as the other preserves order
    return other.preservesOrder();
  }

  @Override
  public UnboundPredicate<T> project(String name, BoundPredicate<T> predicate) {
    return projectStrict(name, predicate);
  }

  @Override
  public UnboundPredicate<T> projectStrict(String name, BoundPredicate<T> predicate) {
    if (predicate.isUnaryPredicate()) {
      return Expressions.predicate(predicate.op(), name);
    } else if (predicate.isLiteralPredicate()) {
      return Expressions.predicate(
          predicate.op(), name, predicate.asLiteralPredicate().literal().value());
    } else if (predicate.isSetPredicate()) {
      return Expressions.predicate(predicate.op(), name, predicate.asSetPredicate().literalSet());
    }
    return null;
  }

  @Override
  public boolean isIdentity() {
    return true;
  }

  @Override
  public boolean equals(Object o) {
    // Can be removed with 2.0.0 deprecation of get(Type)
    if (this == o) {
      return true;
    } else if (o instanceof Identity) {
      return true;
    }
    return false;
  }

  @Override
  public int hashCode() {
    // Can be removed with 2.0.0 deprecation of get(Type)
    return 0;
  }

  @Override
  public String toString() {
    return "identity";
  }

  Object writeReplace() throws ObjectStreamException {
    return SerializationProxies.IdentityTransformProxy.get();
  }
}
